src/abstractdataframe/selection.jl

```diff
@@ -0,0 +1,357 @@
+`old_column` can be a `Symbol` or an integer, ``new_column_name` must be a `Symbol`,
+and `fun` must be a `Function` that is applied row by row to the values of `old_column`.
+
+Column transfomation syntax also supports a short `old_column=>fun` form, in which
+case `new_column_name` is automatically generated by joining `olcd_column` with `fun` name
+with `_`.
+
+If more than one argument is passed then duplicates are accepted except for
+column renaming and transformation operations, where it is not alloweded to rename/transform
+into the same column name.
+For example if `:col` is present in `df` a call `select!(df, :col, :)` is valid
+and moves the column `:col` moved to be the first one in-place.
+
+Note that including the same column several times in the data frame will create aliases.
+
+# Examples
+```jldoctest
+julia> df = DataFrame(a=1:3, b=4:6)
+3×2 DataFrame
+│ Row │ a     │ b     │
+│     │ Int64 │ Int64 │
+├─────┼───────┼───────┤
+│ 1   │ 1     │ 4     │
+│ 2   │ 2     │ 5     │
+│ 3   │ 3     │ 6     │
+
+julia> select!(df, 2)
+3×1 DataFrame
+│ Row │ b     │
+│     │ Int64 │
+├─────┼───────┤
+│ 1   │ 4     │
+│ 2   │ 5     │
+│ 3   │ 6     │
+
+julia> df = DataFrame(a=1:3, b=4:6)
+3×2 DataFrame
+│ Row │ a     │ b     │
+│     │ Int64 │ Int64 │
+├─────┼───────┼───────┤
+│ 1   │ 1     │ 4     │
+│ 2   │ 2     │ 5     │
+│ 3   │ 3     │ 6     │
+
+julia> select!(df, :a=><(1.5)=>:c, :b)
+3×2 DataFrame
+│ Row │ c    │ b     │
+│     │ Bool │ Int64 │
+├─────┼──────┼───────┤
+│ 1   │ 1    │ 4     │
+│ 2   │ 0    │ 5     │
+│ 3   │ 0    │ 6     │
+```
+
+"""
+function select!(df::DataFrame, inds::AbstractVector{Int})
+    if isempty(inds)
+        empty!(_columns(df))
+        empty!(index(df))
+        return df
+    end
+    indmin, indmax = extrema(inds)
+    if indmin < 1
+        throw(ArgumentError("indices must be positive"))
+    end
+    if indmax > ncol(df)
+        throw(ArgumentError("indices must not be greater than number of columns"))
+    end
+    if !allunique(inds)
+        throw(ArgumentError("indices must not contain duplicates"))
+    end
+    copy!(_columns(df), _columns(df)[inds])
+    x = index(df)
+    copy!(_names(x), _names(df)[inds])
+    empty!(x.lookup)
+    for (i, n) in enumerate(x.names)
+        x.lookup[n] = i
+    end
+    return df
+end
+
+select!(df::DataFrame, c::Int) = select!(df, [c])
+select!(df::DataFrame, c::Union{AbstractVector{<:Integer}, AbstractVector{Symbol},
+                                Colon, All, Not, Between, Regex}) =
+    select!(df, index(df)[c])
+
+function select!(df::DataFrame, cs...)
+    newdf = select(df, cs..., copycols=false)
+    copy!(_columns(df), _columns(newdf))
+    x = index(df)
+    copy!(_names(x), _names(newdf))
+    empty!(x.lookup)
+    for (i, n) in enumerate(x.names)
+        x.lookup[n] = i
+    end
+    return df
+end
+
+"""
+    transform!(df::DataFrame, cs...)
+
+The same as [`select!`](@ref) but retains all columns existing in `df`.
```

Comment:
Maybe leave `transform` out for now? If it's just a convenience wrapper around `select`, we can easily add it later once we have settled its design.

Reply:
I am OK to remove it. I added it because I felt people explicitly wanted it (actually I thought that having `select` only is enough, but there were comments that people would prefer to have it as a separate function). What do you think? (and what are the design decisions we have to make with `transform`?)